File: js/longestSubstring.js
// Time complexity: O(n)
// Space complexity: O(n)

function longestSubstring(string) {
  let count = 0;
  let temp = [];

  for (let i = 0; i < string.length; i++) {
    if (!temp.includes(string[i])) {
      temp.push(string[i]);
    } else {
      if (temp.length > count) {
        count = temp.length;
      }
      temp.length = 0;
      temp.push(string[i]);
    }
    if (i == string.length - 1 && count == 0) {
      return temp.length;
    }
  }
  return count;
}

const strings = ["abcabccb", "dorime", "hokpol"];
for (string of strings) {
  console.log(longestSubstring(string));
}

File: kt/longestSubstring.kt
/*kotlin 1.8.0

Time complexity: O(n)
Space complexity: O(n)
 */

fun main() {
  val strings = listOf("abcabccb", "dorime", "hopjol")
  for (string in strings) {
    println(solution(string))
  }
}

fun solution(string: String): Int {
  var count = 0
  val temp: MutableList<Char> = mutableListOf()

  for ((index, ch) in string.withIndex()) {
    if (ch !in temp) {
      temp.add(ch)
    } else {
      if (temp.size > count) {
        count = temp.size
      }
      temp.clear()
      temp.add(ch)
    }
    if (index == string.length - 1 && count == 0) {
      return temp.size
    }
  }
  return count
}
